wrapper/src/main/java/software/amazon/jdbc/plugin/efm2/MonitorImpl.java

```diff
@@ -0,0 +1,418 @@
+
+        ArrayList<Long> processedKeys = new ArrayList<>();
+        this.newContexts.entrySet().stream()
+            // Get entries with key (that is a time in nanos) less or equal than current time.
+            .filter(entry -> entry.getKey() < currentTimeNano)
+            .forEach(entry -> {
+              final Queue<WeakReference<MonitorConnectionContext>> queue = entry.getValue();
+              processedKeys.add(entry.getKey());
+              // Each value of found entry is a queue of monitoring contexts awaiting active monitoring.
+              // Add all contexts to a active monitoring contexts queue.
+              // Ignore contexts that already been disposed.
+              WeakReference<MonitorConnectionContext> contextWeakRef;
+              while ((contextWeakRef = queue.poll()) != null) {
+                MonitorConnectionContext context = contextWeakRef.get();
+                if (context != null && context.isActive()) {
+                  this.activeContexts.add(contextWeakRef);
+                }
+              }
+            });
+        processedKeys.forEach(this.newContexts::remove);
+
+        TimeUnit.SECONDS.sleep(1);
+      }
+    } catch (final InterruptedException intEx) {
+      // do nothing; just exit the thread
+    } catch (final Exception ex) {
+      // this should not be reached; log and exit thread
+      if (LOGGER.isLoggable(Level.FINEST)) {
+        LOGGER.log(
+            Level.FINEST,
+            Messages.get(
+                "MonitorImpl.exceptionDuringMonitoringStop",
+                new Object[]{this.hostSpec.getHost()}),
+            ex); // We want to print full trace stack of the exception.
+      }
+    } finally {
+      telemetryContext.closeContext();
+    }
+  }
+
+  @Override
+  public void run() {
+    final TelemetryContext telemetryContext = telemetryFactory.openTelemetryContext(
+        "monitoring thread", TelemetryTraceLevel.TOP_LEVEL);
+    telemetryContext.setAttribute("url", hostSpec.getUrl());
+
+    try {
+      while (!this.stopped.get()) {
+
+        if (this.activeContexts.isEmpty()) {
+          TimeUnit.MILLISECONDS.sleep(THREAD_SLEEP_MILLIS);
+          continue;
+        }
+
+        final long statusCheckStartTimeNano = this.getCurrentTimeNano();
+        final boolean isValid = this.checkConnectionStatus();
+        final long statusCheckEndTimeNano = this.getCurrentTimeNano();
+
+        this.updateNodeHealthStatus(isValid, statusCheckStartTimeNano, statusCheckEndTimeNano);
+
+        if (this.nodeUnhealthy) {
+          // Kill all connections.
+          WeakReference<MonitorConnectionContext> monitorContextWeakRef;
+          while ((monitorContextWeakRef = this.activeContexts.poll()) != null) {
+            if (this.stopped.get()) {
+              break;
+            }
+            MonitorConnectionContext monitorContext = monitorContextWeakRef.get();
+            if (monitorContext != null) {
+              monitorContext.setNodeUnhealthy(true);
```

Comment:
Do we actually need to pass `true` to this method ? Will we ever explicitly call `ctx.setNodeUnhealthy(false)`?
```suggestion
              monitorContext.setNodeUnhealthy(this.nodeUnhealthy);
```